### ISSUE TITLE
storage: SCATTER hangs during RangeLookup

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#17760]
I'm periodically seeing an `ALTER TABLE ... SCATTER` hang indefinitely. I finally got around to debugging this and the blockage is during a `RangeLookup` call:

```
2017/08/19 01:45:17.902373 | 318.164174 | /cockroach.roachpb.Internal/Batch
           01:45:17.902375 |    .     2 | ... txnID:30da25bb-71c1-461f-9ff3-87a8f18696e9
           01:45:17.902375 |    .     1 | ... txnid:ffb6d7fe-ac2b-4e55-ae1b-22b180db122a
           01:45:17.902379 |    .     4 | ... [n1] 1 RngLookup
           01:45:17.902385 |    .     6 | ... [n1,s1] executing 1 requests
           01:45:17.902390 |    .     5 | ... [n1,s1,r1/1:/{Min-System/}] read-only path
           01:45:17.902395 |    .     5 | ... [n1,s1,r1/1:/{Min-System/}] command queue
           01:45:17.902398 |    .     3 | ... [n1,s1,r1/1:/{Min-System/}] operation accepts inconsistent results
           01:45:17.902401 |    .     3 | ... [n1,s1,r1/1:/{Min-System/}] waiting for read lock
           01:45:17.902407 |    .     7 | ... [n1,s1,r1/1:/{Min-System/}] RangeLookup
           01:45:17.902549 |    .   142 | ... [n1,s1,r1/1:/{Min-System/}] submitting 1 intents to asynchronous processing
```

I see another request is blocked processing the `AdminScatter`. The request log doesn't give all the detail, but it appears a rebalance operation is in progress:

```
01:45:17.853592 | .   292 | ... [s1,r11374/1:/Table/51/1/77{89937…-90749…},n1] r11374: sending batch 1 EndTxn to (n1,s1):1
01:45:17.853606 | .    15 | ... [s1,r11374/1:/Table/51/1/77{89937…-90749…},n1] sending request to local server
01:45:17.856349 | .  2742 | ... [s1,r11374/1:/Table/51/1/77{89937…-90749…},n1] reply error [txn: 30da25bb], EndTransaction [/Local/Range/"\xbb\x89\xfdl\x1bk\x1c\x92\xb8\x05u"/RangeDescriptor,/Min): r11374 was not found
01:45:17.856352 | .     4 | ... [s1,r11374/1:/Table/51/1/77{89937…-90749…},n1] evicting range descriptor on send error and backoff for re-lookup
01:45:17.857875 | .  1522 | ... [s1,r11374/1:/Table/51/1/77{89937…-90749…},n1] evicting cached range descriptor
01:45:17.902318 | . 44443 | ... [s1,r11374/1:/Table/51/1/77{89937…-90749…},n1,range-lookup] r1: sending batch 1 RngLookup to (n1,s1):1
01:45:17.902339 | .    21 | ... [s1,r11374/1:/Table/51/1/77{89937…-90749…},n1,range-lookup] sending request to local server
01:46:17.902644 | 60.000305 | ... kv/dist_sender.go:1156 [s1,r11374/1:/Table/51/1/77{89937…-90749…},n1,range-lookup] have been waiting 1m0s sending RPC to r1 for batch: RangeLookup [/Meta2/Table/51/1/7789937750932194677,/Min)
```

The logs on `n1` show the most recent activity for `r11374`:

```
I170819 01:45:17.308519 342080 storage/replica_command.go:3637  [n1,s1,r11374/1:/Table/51/1/77{89937…-90749…}] change replicas (REMOVE_REPLICA (n5,s5):4): read existing descriptor r11374:/Table/51/1/77{89937750932194677-90749866333379934} [(n1,s1):1, (n6,s6):2, (n3,s3):3, (n5,s5):4, next=5]
```

Going to poke around and try to find out why the `RangeLookup` operation hung. Processing intents asynchronously shouldn't block.

**Comment by petermattis** [comment]
There is 1 goroutine blocked in `processIntentsAsync`:

```
#	0x11f1220	github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).lookupRangeDescriptorInternal+0x640				/go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:343
#	0x11f0b91	github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).LookupRangeDescriptor+0x91					/go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:241
#	0x11e8f3c	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).getDescriptor+0x8c							/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:425
#	0x11f41d6	github.com/cockroachdb/cockroach/pkg/kv.(*RangeIterator).Seek+0x326								/go/src/github.com/cockroachdb/cockroach/pkg/kv/range_iter.go:171
#	0x11eab20	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges+0x290						/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:719
#	0x11e9fb3	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send+0x343								/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:606
#	0x11f8ad0	github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).Send+0x1f0								/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:435
#	0xd7381e	github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).send+0x1fe								/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:554
#	0xd7f5e2	github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).(github.com/cockroachdb/cockroach/pkg/internal/client.send)-fm+0x82	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:491
#	0xd730f2	github.com/cockroachdb/cockroach/pkg/internal/client.sendAndFill+0x102								/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:463
#	0xd7336c	github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Run+0x9c								/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:491
#	0xfae095	github.com/cockroachdb/cockroach/pkg/storage.(*intentResolver).maybePushTransactions+0xcb5					/go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_resolver.go:209
#	0xfaf23f	github.com/cockroachdb/cockroach/pkg/storage.(*intentResolver).processIntents+0x17f						/go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_resolver.go:294
#	0xfaeeee	github.com/cockroachdb/cockroach/pkg/storage.(*intentResolver).processIntentsAsync+0x34e					/go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_resolver.go:272
#	0xff7c04	github.com/cockroachdb/cockroach/pkg/storage.(*Replica).executeReadOnlyBatch+0x7d4						/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2415
```

Looks like we have a deadlock scenario:
1. `DistSender` needs to refresh the range descriptor cache and calls `RangeDescriptorCache.lookupRangeDescriptorInternal` which adds a note that there is a pending lookup operation.
2. The `RangeLookup` finds an intent and calls `IntentResolver.processIntentsAsync`. There are too many async tasks already running, so it uses the current goroutine to process the intent resolution.
3. Resolving the intent involves trying to push the transaction which requires a `DistSender` call which requires looking up the range descriptor we're currently trying to resolve which blocks forever waiting for the prior request to finish.

Note that performing intent resolution on the calling goroutine when there are no async tasks available was a necessary feedback loop. Prior to adding it, we were seeing a spike in intent resolution goroutines under certain scenarios. One possible hacky workaround would be to not perform the intent resolution synchronously for `RangeLookup` operations.

@tschottdorf you're more familiar with this code than I am. Thoughts?

**Comment by bdarnell** [comment]
This also means that we're not enforcing timeouts/cancellation somewhere we ought to. That would have limited the damage from this issue, but it's still better to fix it properly than to rely on timeouts.

It feels like a hack to skip the intent resolution for RangeLookup, but I don't have a better idea right now.

**Comment by petermattis** [comment]
> It feels like a hack to skip the intent resolution for RangeLookup, but I don't have a better idea right now.

Note that we only skip intent resolution for RangeLookup when we can't perform the resolution asynchronously. We have a hardcoded limit of 100 async resolution tasks per node. That might be low, though it is curious why there are so many intent resolution tasks being performed under this workload. That might be worth investigating. 

Interestingly, we do have a timeout (`base.NetworkTimeout`) on the intent resolution task. But we block waiting for the prior range lookup to succeed:

```
kv/range_cache.go:342:
	// Wait for the inflight request.
	res := <-resC
```

Seems like we should definitely change that to:

```
var res Result
select {
  case <-ctx.Done():
    return nil, nil, ctx.Err()
  case res = <-resC
}
```

@nvanbenschoten I'm not terribly familiar with the `RangeDescriptorCache` code. Is there anything more that would be needed here?

@bdarnell Even with proper cancellation/timeout handling in `RangeDescriptorCache`, I think we'd still want to prevent synchronous intent resolution for `RangeLookup` requests in order to avoid a 3s delay for `RangeLookup` ops which could possibly cascade into additional failures.

**Comment by bdarnell** [comment]
> it is curious why there are so many intent resolution tasks being performed under this workload

Range splits immediately trigger RangeKeyMismatchErrors, which can lead to cache invalidations on the clients, concurrent with the split transaction's attempt to clean up its own intents. This effect is amplified by the prefetching of range descriptors, which causes RangeLookups to perform intent resolution on meta records that they aren't directly interested in (this includes both recently-split records whose intents can be resolved, and records for in-progress splits whose PushTransactions will probably fail). 

We have mechanisms to mitigate these issues. For example, RangeKeyMismatchErrors contain the updated range descriptors to insert in the cache. But the concurrency here is subtle and there's a good chance that frequent splits are causing it to break down and invalidate the cache instead of allowing the updated descriptors to be used. 

We probably also want the prefetching mechanism to just skip intents instead of resolving them. Most intents should be resolved by the transaction coordinator that wrote them; the only time it's *necessary* for RangeLookup to resolve intents is to recover from coordinator failure, so we can probably limit that to the primary range addressed by the request. 

> Even with proper cancellation/timeout handling in RangeDescriptorCache, I think we'd still want to prevent synchronous intent resolution for RangeLookup requests in order to avoid a 3s delay for RangeLookup ops which could possibly cascade into additional failures.

Yes, definitely.

**Comment by nvanbenschoten** [comment]
> Seems like we should definitely change that to:
>
> ```
> var res Result
> select {
>   case <-ctx.Done():
>     return nil, nil, ctx.Err()
>   case res = <-resC
> }
> ```

That should absolutely be changed to what you suggest, which should be sufficient. Note that right now a context cancellation for the "leader" request (the request that comes first which all subsequent lookups for the same range join onto) will work properly and will be propagated to all "follower" requests. However, without your proposed change, it looks like a "follower" context cancellation will be ignored indefinitely.

> But the concurrency here is subtle and there's a good chance that frequent splits are causing it to break down and invalidate the cache instead of allowing the updated descriptors to be used.

I've always suspected that repeated splits would be problematic given our current approach of coalescing all lookups for keys within the bounds of a previously split descriptor onto a single `RangeLookup` request. When a split occurs, we group all requests for keys within the evicted descriptor together, understanding that the result of the lookup may not satisfy requests that should have gone to the other side of the split. This means that any requests for keys to the other side of the split from the one they piggybacked onto will find that the [lookup missed](https://github.com/cockroachdb/cockroach/blob/2e76b6e567311330041c97e2f564ece8485feefd/pkg/kv/range_cache.go#L353), and they will then perform the lookup again to find their correct descriptor. With this approach, a problem could clearly arise if the `RangeLookup` requests were performed at any less than double the frequency of range splits, because more and more requests would be incorrect grouped together. One way to counter this could be to stop attempting to coalesce lookups after a certain number of lookup misses for a given key.

**Comment by petermattis** [comment]
> That should absolutely be changed to what you suggest, which should be sufficient.

Can I hand this off to you (along with adding a test)? I'll still tackle the `processIntentAsync` deadlock which I need to figure out how to reproduce with a test. 

> With this approach, a problem could clearly arise if the RangeLookup requests were performed at any less than double the frequency of range splits, because more and more requests would be incorrect grouped together.

There will be some natural rate limiting for user-initiated splits because those `AdminSplit` requests will need the previous `RangeLookup` to succeed in order to direct the new `AdminSplit` request to the correct location.

**Comment by nvanbenschoten** [comment]
@petermattis yes I can take that. See #17795. We will probably also want to add logic for follower requests to retry `RangeCache` lookups when a leader request has a context cancellation but the followers' contexts are still valid.

> There will be some natural rate limiting for user-initiated splits because those AdminSplit requests will need the previous RangeLookup to succeed in order to direct the new AdminSplit request to the correct location.

That's a good point. I guess my scenario would only be possible if the `AdminSplits` went through a different `DistSender` which for some reason was able to perform `RangeLookups` at a much faster rate in order to facilitate the `AdminSplits`.

**Comment by nvanbenschoten** [comment]
> It feels like a hack to skip the intent resolution for RangeLookup, but I don't have a better idea right now.

I realized that the test created for this deadlock has been passing even though new style RangeLookups (simple `ScanRequests`) aren't being caught by this check. I reworked the test to actually use the RangeDescriptorCache and it's now failing as expected. 

We could add more logic to detect these new RangeLookups as well to determine whether to allow sync intent resolution, but I agree that that feels hacky. What about disallowing sync intent resolution on any INCONSISTENT or READ_UNCOMMITTED (see https://github.com/cockroachdb/cockroach/pull/21276) scans? Not only would that solve this deadlock, but it also feels in-line with the nature of these inconsistent requests.
